File: app/src/main/java/aws/apps/usbDeviceEnumerator/ui/usbinfo/fragments/base/DataFetcher.java
package aws.apps.usbDeviceEnumerator.ui.usbinfo.fragments.base;

import android.graphics.Bitmap;
import android.text.TextUtils;

import aws.apps.usbDeviceEnumerator.data.DataProviderCompanyInfo;
import aws.apps.usbDeviceEnumerator.data.DataProviderCompanyLogo;
import aws.apps.usbDeviceEnumerator.data.DataProviderUsbInfo;

/*package*/ class DataFetcher {

    private final DataProviderCompanyInfo dbComp;
    private final DataProviderUsbInfo dbUsb;
    private final DataProviderCompanyLogo zipComp;

    public DataFetcher(DataProviderCompanyInfo dbComp, DataProviderUsbInfo dbUsb, DataProviderCompanyLogo zipComp) {
        this.dbComp = dbComp;
        this.dbUsb = dbUsb;
        this.zipComp = zipComp;
    }

    public void fetchData(final String vid,
                          final String pid,
                          final String reportedVendorName,
                          final Callback callback) {

        final Runnable runnable = () -> {
            final String vendorFromDb;
            final String productFromDb;
            final Bitmap bitmap;

            if (dbUsb.isDataAvailable()) {
                vendorFromDb = dbUsb.getVendorName(vid);
                productFromDb = dbUsb.getProductName(vid, pid);

                if (dbComp.isDataAvailable()) {
                    final String searchFor;

                    if (!TextUtils.isEmpty(vendorFromDb)) {
                        searchFor = vendorFromDb;
                    } else {
                        searchFor = reportedVendorName;
                    }

                    final String logo = dbComp.getLogoName(searchFor);
                    bitmap = zipComp.getLogoBitmap(logo);
                } else {
                    bitmap = null;
                }
            } else {
                vendorFromDb = null;
                productFromDb = null;
                bitmap = null;
            }

            callback.onSuccess(vendorFromDb, productFromDb, bitmap);

        };

        final Thread thread = new Thread(runnable);
        thread.start();
    }

    public interface Callback {

        void onSuccess(String vendorFromDb, String productFromDb, Bitmap bitmap);

    }

}

File: app/src/main/java/aws/apps/usbDeviceEnumerator/ui/usbinfo/fragments/base/DataFetcher.kt
package aws.apps.usbDeviceEnumerator.ui.usbinfo.fragments.base

import android.content.res.Resources
import android.graphics.Bitmap
import uk.co.alt236.usbinfo.database.model.DbResult
import uk.co.alt236.usbinfo.database.providers.DataProviderCompanyInfo
import uk.co.alt236.usbinfo.database.providers.DataProviderCompanyLogo
import uk.co.alt236.usbinfo.database.providers.DataProviderUsbInfo

class DataFetcher(
    private val resources: Resources,
    private val dbComp: DataProviderCompanyInfo,
    private val dbUsb: DataProviderUsbInfo,
    private val zipComp: DataProviderCompanyLogo
) {
    fun fetchData(
        vid: String,
        pid: String,
        reportedVendorName: String?,
        callback: Callback
    ) {
        val runnable = Runnable {
            val vendorFromDb = dbUsb.getVendorName(vid)
            val productFromDb = dbUsb.getProductName(vid, pid)
            val bitmap: Bitmap?

            val logoVendorName = if (!vendorFromDb.getValueOrNull().isNullOrEmpty()) {
                vendorFromDb.getValueOrNull()
            } else if (!reportedVendorName.isNullOrEmpty()) {
                reportedVendorName
            } else {
                null
            }

            bitmap = getLogo(logoVendorName)
            callback.onSuccess(
                vendorFromDb.getUiString(),
                productFromDb.getUiString(),
                bitmap
            )
        }

        val thread = Thread(runnable)
        thread.start()
    }


    private fun getLogo(vendorName: String?): Bitmap? = if (vendorName.isNullOrEmpty()) {
        null
    } else {
        val logoName: String? = dbComp.getLogoName(vendorName).getValueOrNull()
        zipComp.getLogoBitmap(logoName)
    }


    private fun DbResult<String?>.getUiString(): String {
        return when (this) {
            DbResult.DbFailedToOpen -> "Error: Failed to open DB"
            DbResult.DbNotPresent -> "DB not present"
            is DbResult.ErrorGeneric -> "Error: ${this.error.message ?: this.error::class.java.simpleName}"
            is DbResult.Success -> this.result ?: "Not found"
        }
    }

    interface Callback {
        fun onSuccess(vendorFromDb: String?, productFromDb: String?, bitmap: Bitmap?)
    }
}
